File: python/zmq/zguide/clonesrv1.py
# clone server model one
import random, time, zmq
from kvsimple import KVMsg

def main():
	# prepare our context and publisher socket
	ctx = zmq.Context()
	publisher = ctx.socket(zmq.PUB)
	publisher.bind('tcp://*:5556')
	time.sleep(0.2)
	random.seed(time.time())

	kvmap = {}
	sequence = 0

	try:
		while True:
			# distribute as key-value message
			sequence += 1
			kvmsg = KVMsg(sequence)
			kvmsg.key = '%d' % random.randint(1, 10000)
			kvmsg.body = '%d' % random.randint(1, 1000000)
			kvmsg.send(publisher)
			kvmsg.store(kvmap)
	except KeyboardInterrupt:
		print ' Interrupted\n%d messages out' % sequence

if __name__ == '__main__':
	main()

File: zmq3/clonesrv1.cpp
// clone server model one
#include <czmq.h>
#include "kvsimple.h"

int main(int argc, char * argv[])
{
	zctx_t * ctx = zctx_new();
	void * publisher = zsocket_new(ctx, ZMQ_PUB);
	zsocket_bind(publisher, "tcp://*:5556");
	zclock_sleep(200);

	zhash_t * kvmap = zhash_new();
	int64_t sequence = 0;
	srandom((unsigned)time(NULL));

	while (!zctx_interrupted)
	{
		// distribute as key-value message
		kvmsg_t * kvmsg = kvmsg_new(++sequence);
		kvmsg_fmt_key(kvmsg, "%d", randof(10000));
		kvmsg_fmt_body(kvmsg, "%d", randof(1000000));
		kvmsg_send(kvmsg, publisher);
		kvmsg_store(&kvmsg, kvmap);
	}

	printf(" Interrupted\n%d messages out\n", (int)sequence);
	zhash_destroy(&kvmap);
	zctx_destroy(&ctx);

	return 0;
}
